tools/bug-report/pkg/cluster/cluster.go

```diff
@@ -0,0 +1,166 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"istio.io/pkg/log"
+)
+
+type ResourceType int
+
+const (
+	Namespace ResourceType = iota
+	Deployment
+	Pod
+	Label
+	Annotation
+	Container
+)
+
+func ParsePath(path string) (namespace string, deployment, pod string, container string, err error) {
+	pv := strings.Split(path, "/")
+	if len(pv) != 4 {
+		return "", "", "", "", fmt.Errorf("bad path %s, must be namespace/pod/container", path)
+	}
+	return pv[0], pv[1], pv[2], pv[3], nil
+}
+
+// GetClusterResources returns cluster resources for the given REST config and k8s Clientset.
+func GetClusterResources(ctx context.Context, clientset *kubernetes.Clientset) (*Resources, error) {
+	var errs []string
+	out := &Resources{
+		Labels:      make(map[string]map[string]string),
+		Annotations: make(map[string]map[string]string),
+		Pod:         make(map[string]*corev1.Pod),
+	}
+	namespaces, err := clientset.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	for _, ns := range namespaces.Items {
+		pods, err := clientset.CoreV1().Pods(ns.Name).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return nil, err
+		}
+		replicasets, err := clientset.AppsV1().ReplicaSets(ns.Name).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return nil, err
+		}
+		for _, p := range pods.Items {
+			deployment, err := getOwnerDeployment(&p, replicasets.Items)
+			if err != nil {
+				errs = append(errs, err.Error())
+				continue
+			}
+			for _, c := range p.Spec.Containers {
+				out.insertContainer(ns.Name, deployment, p.Name, c.Name)
+			}
+			out.Labels[p.Name] = p.Labels
+			out.Annotations[p.Name] = p.Annotations
+			out.Pod[p.Name] = &p
+		}
+	}
+	if len(errs) != 0 {
+		log.Warna(strings.Join(errs, "\n"))
+	}
+	return out, nil
+}
+
+// Resources defines a tree of cluster resource names.
+type Resources struct {
+	// Root is the first level in the cluster resource hierarchy.
+	// Each level in the hierarchy is a map[string]interface{} to the next level.
+	// The levels are: namespaces/deployments/pods/containers.
+	Root map[string]interface{}
+	// Labels maps a pod name to a map of labels key-values.
+	Labels map[string]map[string]string
+	// Annotations maps a pod name to a map of annotation key-values.
+	Annotations map[string]map[string]string
+	// Pod maps a pod name to its Pod info.
+	Pod map[string]*corev1.Pod
```

Comment:
I am not really sure whether I understand correctly what the Resource struct represents, looks like it is for pod level report? Can we rename it to something more specific since the name is too general.

Reply:
it's meant to be used together with the package name i.e. cluster.Resources.